### PR TITLE
refactor(ts): update index with new type

### PR DESCRIPTION
**Pull request by Haroenv**
**Summary**

updates the index widget

**Result**

because the types don't match, Index no longer is a Widget, which has impact in addWidget / removeWidget / getWidget

### CODE REVIEW

File: src/lib/InstantSearch.ts [hunk 1]
```diff
@@ -339,7 +339,7 @@ See ${createDocumentationLink({
    * Widgets can be added either before or after InstantSearch has started.
    * @param widgets The array of widgets to add to InstantSearch.
    */
-  public addWidgets(widgets: Widget[]) {
+  public addWidgets(widgets: Array<Widget | Index>) {
```
Comment (eunjae-lee) [inline src/lib/InstantSearch.ts:342]: This gives me the impression that `index` is not a widget at all, whereas in fact it's a sort of special widget.

```suggestion
  public addWidgets(widgets: Array<Widget | IndexWidget>) {
```

I think it aligns better with other widgets like `HitsWidget, RefinementListWidget, ...` and also with `isIndexWidget()` function.

**Review by eunjae-lee — APPROVED** [review]
Looks clear to me.

File: src/lib/InstantSearch.ts (hunk 1 above)
Comment (Haroenv) [reply src/lib/InstantSearch.ts:342]: makes sense, let's do this while it's not yet a breaking change

File: src/widgets/index/index.ts [hunk 2]
```diff
@@ -641,7 +659,11 @@ const index = (props: IndexProps): Index => {
           // `dispose` because the index is removed. We can't call `removeWidgets`
           // because we want to keep the widgets on the instance, to allow idempotent
           // operations on `add` & `remove`.
-          widget.dispose({ helper: helper!, state: helper!.state });
+          widget.dispose({
+            helper: helper!,
+            state: helper!.state,
+            parent: this,
+          });
```
Comment (Haroenv) [inline src/widgets/index/index.ts:666]:
```suggestion
          widget.dispose({ helper: helper!, state: helper!.state });
```

File: src/widgets/index/index.ts [hunk 3]
```diff
@@ -351,7 +365,11 @@ const index = (props: IndexProps): Index => {
       if (localInstantSearchInstance && Boolean(widgets.length)) {
         const nextState = widgets.reduce((state, widget) => {
           // the `dispose` method exists at this point we already assert it
-          const next = widget.dispose!({ helper: helper!, state });
+          const next = widget.dispose!({
+            helper: helper!,
+            state,
+            parent: this,
+          });
```
Comment (Haroenv) [inline src/widgets/index/index.ts:372]:
```suggestion
          const next = widget.dispose!({ helper: helper!, state });
```

**Review by shortcuts — APPROVED** [review]

**Pull request merged**